NAT address discovery

tzaeschke opened the pull request.

Commit:
NAT

Commit:
NAT

Commit:
NAT

Commit:
NAT

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+    NATed IP/port).
+
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: Conceptually a bit of a hack. More standardization effort? How do we solve authentication?
```
Another clear disadvantage of this approach: The BR would need to check every outbound packet as part of the fast path.

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+-  Reliability: The border router is almost guaranteed to see the correct IP/port on the NAT, especially when using the
+   same port for STUN traffic and routing traffic. All other approaches rely on the leniency of the NAT to use the same
+   port even if the STUN server and border router have different ports or even IPs.
+-  Time to rollout: changing the border routers should be much easier and faster than getting router vendors or NAT
```
Remove "router vendors or"?

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+
+-  The SRC address/port is updated by the border router to reflect what the border router sees as source address.
+   Problem:
+   -  Complicates cryptographically protecting the header if the header must be modifiable by border routers
```
Refer to SPAO [https://docs.scion.org/en/latest/protocols/authenticator-option.html]

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+-  The SRC address/port is updated by the border router to reflect what the border router sees as source address.
+   Problem:
+   -  Complicates cryptographically protecting the header if the header must be modifiable by border routers
+   -  A spoofed IP causes the border router to route traffic to an unsuspecting target
```
This is always true, right? Maybe remove the entire sentence?

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+   Problem:
+   -  Complicates cryptographically protecting the header if the header must be modifiable by border routers
+   -  A spoofed IP causes the border router to route traffic to an unsuspecting target
+-  The SRC address/port is updated by the NAT. This is similar to having the border router update tSRC IP/port.
```
"the SRC" instead of "tSRC"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+   -  Complicates cryptographically protecting the header if the header must be modifiable by border routers
+   -  A spoofed IP causes the border router to route traffic to an unsuspecting target
+-  The SRC address/port is updated by the NAT. This is similar to having the border router update tSRC IP/port.
+   - Complicates cryptographically protecting the header if the header must be modifiable by border routers
```
> modifiable by border router

"modifiable by NATs"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+   -  A spoofed IP causes the border router to route traffic to an unsuspecting target
+-  The SRC address/port is updated by the NAT. This is similar to having the border router update tSRC IP/port.
+   - Complicates cryptographically protecting the header if the header must be modifiable by border routers
+-  Use STUN servers. This is a possibility, but adds setup complexity and may not work in all cases. Every subnet
```
> Use STUN servers.

Try to make this more precise: "Use separate STUN servers" or "Use STUN servers other than the BRs", ...

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+   - Complicates cryptographically protecting the header if the header must be modifiable by border routers
+-  Use STUN servers. This is a possibility, but adds setup complexity and may not work in all cases. Every subnet
+   of an AS that has a border router must also have a STUN server. Moreover, if the STUN server uses a different IP
+   (or port) than the border router, then the NAT may decide to use a different port when connecting to it, i.e. the
```
> a different  port when connecting to it

"a different port mapping"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+   STUN server may not see the same IP/port tuple on the NAT that the border router sees. Disadvantages:
+
+   -  This approach may be be problematic with sensitive NATs.
+   -  We need to somehow standardize the STUN IP/port and/or communicate it to endhosts, e.g. via the topo file or
```
"topology.json" instead of "topo"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+   -  This approach may be be problematic with sensitive NATs.
+   -  We need to somehow standardize the STUN IP/port and/or communicate it to endhosts, e.g. via the topo file or
+      its successor.
+-  Remove all NATs and use IPv6 instead. This is technically possible but unlikely to happen anytime soon.
```
Maybe add something along the lines "especially because scarcity of IPv4 addresses is not the only reason why NATs are deployed".

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+Transition
+----------
+
+- An "old" client without expectation on STUN support on the router would simply not use this feature. No problem here.
```
> No problem here.

"No additional problem here."?

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+- An "old" client without expectation on STUN support on the router would simply not use this feature. No problem here.
+- A "new" client sending a STUN request to an "old" border router should simply fail because the router should simply
+  drop a packet that it cannot process.
+  The client should then time out and report that the external NAT address cloud not be established.
```
> cloud

"could"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+- An "old" client without expectation on STUN support on the router would simply not use this feature. No problem here.
+- A "new" client sending a STUN request to an "old" border router should simply fail because the router should simply
+  drop a packet that it cannot process.
+  The client should then time out and report that the external NAT address cloud not be established.
```
Wouldn't it be better in this case to maybe log the above fact but instead of failing to optimistically assume that no NAT is deployed?

tzaeschke replied:
I added that as an alternative with `TBD`. You are probably right, but I think some people may prefer erroring out because it clearly indicates an outdated border router installation.

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+  drop a packet that it cannot process.
+  The client should then time out and report that the external NAT address cloud not be established.
+
+**TODO** How can we avoid this failure in cases where there is no NAT? Can we get the BR version or CS version?
+If they are outdated then the client can try without NAT resolution which may simply work if there is not
+NAT or fail if there is one.
```
See above.

marcfrei reviewed (commented):
Thank you very much Til! Here the comments from a first pass through the document.

marcfrei commented:
Re the title of the PR: "NAT" instead of "Nat"

tzaeschke commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,143 @@
+[Introduction of the topic, description of the problem being solved.]
+
+After a client endhost sends a request to a server over SCION, in order to return a packet response to the client,
+the border router in the client's AS uses the SRC address from the SCION header and the port of the UDP payload to reach
```
Well, it uses the SRC address of the packet it received as DST address. I tried to clarify. Is it better now?

Commit:
Feedback Marc

tzaeschke commented:
@marcfrei Thanks for the feedback! It should all be fixed now, except maybe the SRC/DST address issue?

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+   -  Complicates cryptographically protecting the header if the header must be modifiable by border routers,
+      e.g. for `SPAO <https://docs.scion.org/en/latest/protocols/authenticator-option.html>`_.
+-  The SRC address/port is updated by the NAT. This is similar to having the border router update SRC IP/port.
+   - Complicates cryptographically protecting the header if the header must be modifiable by NATs.
```
Maybe also mention here again that this would in addition require cooperation of the NAT suppliers to include and roll-out SCION support to all their devices.

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+In case the detector shares and IP and possibly even port (or uses a fixed port) with the border router, this
+discovery is straight forward.
+If the port is flexible or the detector runs on a different IP then we need to find a different solution, probably
+as an extension to the discover/bootstrapping service.
```
"discover/bootstrapping service" > "bootstrapping service"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+
+   -  This approach may be be problematic with sensitive NATs.
+   -  We need to somehow standardize the STUN IP/port and/or communicate it to endhosts, e.g. via the topology.json file
+      or its successor.
```
"its successor" > "the bootstrapping service"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+Transition
+----------
+
+- An "old" client without expectation on STUN support on the router would simply not use this feature.
```
"STUN support" > "NAT IP/port discovery support"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+
+- An "old" client without expectation on STUN support on the router would simply not use this feature.
+  No additional problem here.
+- A "new" client sending a STUN request to an "old" border router should simply fail because the router should simply
```
"STUN request" > "NAT IP/port discovery request"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+  The client should then time out and report that the external NAT address could not be established. Instead of timing
+  out it could also optimistically assume that no NAT is involved. -> TBD
+
+**TODO** How can we avoid this failure in cases where there is no NAT? Can we get the BR version or CS version?
+If they are outdated then the client can try without NAT resolution which may simply work if there is not
+NAT or fail if there is one. Or the client could optimistically assume that no NAT is present.
```
Do we really need this? Otherwise, remove this section.

marcfrei reviewed (commented):
Three problems/points, that we still have to address/mention:

(1.) What if the packet doesn't return via the same AS-local, first-hop/last-hop BR and the NAT is sensitive to the source IP of incoming packets? This is possible because servers don't have to simply reverse the path of request packets, they are allowed to do their own path discovery and selection.

(2.) What if the the port mapping at the NAT doesn't result in a L4 port that is within the "dispatched_ports" range (https://docs.scion.org/en/latest/dev/design/router-port-dispatch.html)? The last-hop BR would then choose the fixed end-host data port (default 30041) as the underlay UDP/IP destination port and NAT mapping would fail. Contrary to point (1.) this problem would at least be detectable by the endhost before based on the proposed NAT IP/port discovery mechanism.

(3.) Is there a simple way to find out how sensitive NATs in practice actually are, maybe via the links to Tailscale in the issue #4517?

tzaeschke commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+[Introduction of the topic, description of the problem being solved.]
+
+After a client endhost sends a request to a server over SCION, in order to return a packet response to the client,
+the border router in the client's AS uses the SRC address from the received SCION header and the port of the UDP payload
```
I rephrased the whole paragraph, could you have another look?

tzaeschke commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+Another complication is that some NATs may change their port mapping port if a client endhosts connects to a new
+remote IP or port or uses a different local port.
+
+Therefore, it is desirable for the detector not only be in the same AS and same subnet as the BR, but ideally on the
```
Move introduction to 1st paraggraph in this section

tzaeschke commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,148 @@
+Another complication is that some NATs may change their port mapping port if a client endhosts connects to a new
+remote IP or port or uses a different local port.
+
+Therefore, it is desirable for the detector not only be in the same AS and same subnet as the BR, but ideally on the
```
I think I would like to keep that in to emphasize that there is a difference, an AS may have multiple subnets. Also, in theory, there could be multiple AS in one subnet, or not?

tzaeschke reviewed (commented):
Thanks! I took over most proposed changes, a few are still open.

Commit:
Feedback Marc round 2

tzaeschke commented:
> Three problems/points, that we still have to address/mention:
> 
> (1.) What if the packet doesn't return via the same AS-local, first-hop/last-hop BR and the NAT is sensitive to the source IP of incoming packets? This is possible because servers don't have to simply reverse the path of request packets, they are allowed to do their own path discovery and selection.
> 
> (2.) What if the the port mapping at the NAT doesn't result in a L4 port that is within the "dispatched_ports" range (https://docs.scion.org/en/latest/dev/design/router-port-dispatch.html)? The last-hop BR would then choose the fixed end-host data port (default 30041) as the underlay UDP/IP destination port and NAT mapping would fail. Contrary to point (1.) this problem would at least be detectable by the endhost before based on the proposed NAT IP/port discovery mechanism.
> 
> (3.) Is there a simple way to find out how sensitive NATs in practice actually are, maybe via the links to Tailscale in the issue [scionproto#4517](https://github.com/scionproto/scion/issues/4517)?

(1.) I added a paragraph to the "Compatibility" section: 
`However, the proposal requires packets that come from a server to arrive through the same BR that was used for an
outgoing request. This currently seems to be the default behavior of servers, but it is not a behavior required
by the current standard.`

(2.) I guess this is an open problem. However, I think this port range is meant to disappear, so this may be resolved before our NAT proposal is being rolled out. I added your question to "Compatibility"

(3.) I am not sure how we can find out how sensitive they are in practice. I am also not sure if that would really help. On the lower end, our approach is similar to STUN, so if that works, than our should work as well (unless NATs recognize STUN request and behave differently?). On the upper end, I think we should aim for supporting maximally fickle NATs?

Commit:
Feedback Marc round 2

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+
+One complication is that the local AS of the sender may be split into different subnets and that border routers
+are not all in the same subnet.
+Another complication is that some NATs may change their port mapping port if a client endhost connects to a new
```
NIT: "port mapping port" > "port mapping" (is probably clear enough)

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+    to change the payload so it contains the IP header of the offending packet (and the IP header should contain the
+    NATed IP/port).
+
+    -  Advantages: One less dependency on an external library and protocol
```
Plus: one roundtrip less in case an endhost doesn't sit behind a NAT or similar.

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+The main reasons for integrating the functionality with the BR are:
+
+-  Reliability: The border router is almost guaranteed to see the correct IP/port on the NAT, especially when using the
+   same port for STUN traffic and routing traffic. All other approaches rely on the leniency of the NAT to use the same
```
"STUN traffic" > "NAT detection traffic"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+
+-  Reliability: The border router is almost guaranteed to see the correct IP/port on the NAT, especially when using the
+   same port for STUN traffic and routing traffic. All other approaches rely on the leniency of the NAT to use the same
+   port even if the STUN server and border router have different ports or even IPs.
```
"STUN server" > "NAT detector"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+-  The SRC address/port is updated by the NAT. This is similar to having the border router update SRC IP/port.
+
+   - Complicates cryptographically protecting the header if the header must be modifiable by NATs.
+   - Requires cooperation of NAT suppliers to include and roll-out SCION support to all their devices.
```
NIT: "roll-out" > "roll out"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+   - Requires cooperation of NAT suppliers to include and roll-out SCION support to all their devices.
+
+-  Use separate STUN servers. This is a possibility, but adds setup complexity and may not work in all cases. Every
+   subnet of an AS that has a border router must also have a STUN server. Moreover, if the STUN server uses a different
```
"must also have" > "would also need"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+
+-  Use separate STUN servers. This is a possibility, but adds setup complexity and may not work in all cases. Every
+   subnet of an AS that has a border router must also have a STUN server. Moreover, if the STUN server uses a different
+   IP (or port) than the border router, then the NAT may decide to use a different port mapping when connecting to it,
```
Remove "when connecting to it"

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+  The last-hop BR would then choose the fixed end-host data port (default 30041) as the underlay UDP/IP destination port
+  and NAT mapping would fail. Contrary to point (1.) this problem would at least be detectable by the endhost before
+  based on the proposed NAT IP/port discovery mechanism.
+
```
One additional point to check: all client libraries (snet/pan, jpan) have to accept incoming packets where the underlay UDP port does not match with the L4 port, i.e., the underlay port could be the local receiver port whereas the L4 port will be the NATed port.

marcfrei commented on doc/dev/design/NAT-address-discovery.rst:
```diff
@@ -0,0 +1,163 @@
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: Conceptually a bit of a hack. The BR would need to check every outbound packet as part of the fast
+       path. More standardization effort? How do we solve authentication?
+
```
To avoid potential misunderstandings make it here explicit, that this proposal only addresses the problem of SCION clients behind NATs. It is not the goal to also support NAT traversal techniques such as NAT hole punching SCION-based servers and peer-to-peer scenarios.

marcfrei reviewed (approved):
I'm very happy with the current state of the document. Thank you so much, Til!

I just added a few more comments. If you decide to address them, please feel free to go ahead and open the PR against scionproto afterwards.

Commit:
Feedback Marc round 3

Commit:
Removed (some) markers

Commit:
Implementation
